### PR TITLE
feat: do not re-install if source does not change

### PR DESCRIPTION
**Pull request by tdejager**
This should ensure that a build/rebuild is only triggered when the source is changed.

The final todo, I would say is to move the merge the `reinstall_globs.rs` file and the new struct that have been created in `environment.rs`.

### CODE REVIEW

**Comment by baszalmstra** [comment]
Soooo I have refactored almost everything. I had a long discussion with @wolfv today and we decided that instead of storing the caching information in the prefix itself we would store it in a cache together with the built artifact. This is then very similar to what we do with the metadata basically.

**Comment by tdejager** [comment]
Could you elaborate why :)?

**Comment by tdejager** [comment]
Yes that makes sense! Thanks for explaining!

File: src/build/cache/build_cache.rs [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    /// An additional directory is created by this cache inside the passed root
+    /// which includes a version number. This is to ensure that the cache is
+    /// never corrupted if the format changes in the future.
+    pub fn new(root: PathBuf) -> Self {
+        Self {
+            root: root.join("source-builds-v0"),
+        }
+    }
+
+    pub async fn entry(
```
Comment (tdejager) [inline src/build/cache/build_cache.rs:93]: Docs.

File: crates/pixi_record/src/pinned_source.rs [hunk 2]
```diff
@@ -63,13 +70,43 @@ impl PinnedSourceSpec {
         }
     }
 
+    #[allow(clippy::result_large_err)]
+    pub fn into_mutable(self) -> Result<MutablePinnedSourceSpec, PinnedSourceSpec> {
```
Comment (tdejager) [inline crates/pixi_record/src/pinned_source.rs:74]: Docs

File: crates/pixi_record/src/pinned_source.rs [hunk 3]
```diff
@@ -63,13 +70,43 @@ impl PinnedSourceSpec {
...
+
     /// Returns true if the pinned source will never change. This can be useful
     /// for caching purposes.
     pub fn is_immutable(&self) -> bool {
         !matches!(self, PinnedSourceSpec::Path(_))
     }
 }
 
+impl MutablePinnedSourceSpec {
+    pub fn as_path(&self) -> Option<&PinnedPathSpec> {
```
Comment (tdejager) [inline crates/pixi_record/src/pinned_source.rs:89]: Docs

File: crates/pixi_record/src/pinned_source.rs [hunk 4]
```diff
@@ -63,13 +70,43 @@ impl PinnedSourceSpec {
...
+        match self {
+            MutablePinnedSourceSpec::Path(spec) => Some(spec),
+        }
+    }
+
+    pub fn into_path(self) -> Option<PinnedPathSpec> {
```
Comment (tdejager) [inline crates/pixi_record/src/pinned_source.rs:95]: Docs

File: src/build/mod.rs [hunk 5]
```diff
@@ -125,21 +129,84 @@ impl BuildContext {
...
+                        .ok()
+                        .unwrap_or(target_platform),
+                    name: source_spec.package_record.name.as_normalized().to_string(),
+                    version: source_spec.package_record.version.to_string(),
+                    build: source_spec.package_record.build.clone(),
+                },
+            )
+            .await?;
+
+        if let Some(build) = cached_build {
```
Comment (tdejager) [inline src/build/mod.rs:160]: Can we group the logging for debugging purposes, you can then see in one glance what was rebuild and why. I noticed this already being annoying when waiting of source packages that take long to build. This way the user can cancel the process before the builds. Can also be a future improvement in a PR.

**Review by tdejager — COMMENTED** [review]
Nice, missing some docs mainly.

File: src/build/mod.rs (hunk 5 above)
Comment (tdejager) [reply src/build/mod.rs:160]: Decided to do this in a future PR.

**Pull request merged**